js-release/src/index.ts

```diff
@@ -0,0 +1,127 @@
+import { exec } from "child_process";
+import readdirp from "readdirp";
+import * as core from "@actions/core";
+import { Storage } from "@google-cloud/storage";
+import setupGCloud from "./setup-gcloud/setup-gcloud";
+
+// config
+const BUCKET = "manifold-js";
+const URL_MAP = "manifold-js-cdn-urlmap";
+
+/**
+ * Publish to npm
+ */
+async function npmPublish({
+  npmDistTag,
+  npmToken,
+  workspace,
+}: {
+  npmDistTag: string;
+  npmToken: string;
+  workspace: string;
+}) {
+  await exec(`echo "//registry.npmjs.org/:_authToken=${npmToken}" > ~/.npmrc`);
+  await exec(`npm publish ${workspace} --tag ${npmDistTag}`);
+}
+
+/**
+ * Publish to Manifold CDN
+ */
+async function cdnPublish({
+  packageName,
+  semver,
+  serviceAccountKey,
+  workspace,
+}: {
+  packageName: string;
+  semver: string;
+  serviceAccountKey: string;
+  workspace: string;
+}) {
+  // recursively read workspace dir
+  const files = await readdirp.promise(workspace);
+
+  // upload files to CDN
+  const storage = new Storage({ keyFile: serviceAccountKey });
+  files.forEach((file) => {
+    storage.bucket(BUCKET).upload(file.fullPath, {
+      destination: `${packageName}@${semver}/${file.path}`,
+    });
+  });
+
+  // update URL Map
+  await setupGCloud();
+  // latest
+  await exec(
+    `gcloud compute url-maps remove-path-matcher ${URL_MAP} --path-matcher-name=latest`
+  );
+  await exec(
+    `gcloud compute url-maps add-path-matcher ${URL_MAP} --path-matcher-name=latest --path-rules=/${packageName}=/${packageName}@${semver}`
+  );
+  // major
+  const major = semver.split(".")[0];
+  try {
+    await exec(
+      `gcloud compute url-maps remove-path-matcher ${URL_MAP} --path-matcher-name=v${major}`
+    );
+  } catch (err) {
+    // new path matcher name; ignore
```

Comment:
These are probably good oppotunity to crash with the error if we want to be sure our paths are always up to date.
We may want to have a clean up function that reverts our changes though.
That way the github action fails and lets us know.
Ideally we'd notify slack or something though so we know, i find failed GH actions on master are pretty invisible, we can do that outside of the action code with a github notification or soemthing.

Reply:
Yeah that’s a good point. I was originally thinking about the first run where the removal would fail, but you’re right—beyond that we should surface errors

Reply:
🤔 Also I think there’s a different way to surface errors back to GitHub Actions, that requires more than a `throw`. I should look that up.